Add a filter to allow getting more values from block_field()

kienstra opened the pull request:
* An alternative to #373 
* This will make it possible for `block_field()` to return or echo additional block attributes that another plugin might have added, like in this [example usage](https://github.com/phpbits/block-options/blob/57a752f703590ea719bb9c85fad1f3a0e97b1fa0/includes/helper.php#L10-L17) of this filter. 
* It's possible to add block attributes [by filtering](https://github.com/phpbits/block-options/blob/9c6aadfc8d7a5153d17acb0bb3630296cfa21c8b/src/extensions/attributes/index.js#L28) the JavaScript function `registerBlockType()`. For example, Editorskit [adds an editorskit object](https://github.com/phpbits/block-options/blob/9c6aadfc8d7a5153d17acb0bb3630296cfa21c8b/src/extensions/attributes/index.js#L28-L43) to the block attributes, and uses it [in the Inspector Controls](https://github.com/phpbits/block-options/blob/aad4e9aa0f49ba29d226ecbf7357f60e7863ed3a/src/extensions/advanced-controls/index.js#L101).
* But before, it wasn't possible to output these attributes with `block_field()`. That only allowed registered Block Lab fields and `className`.

kienstra committed:
Add a filter to allow getting more values from block_field()

Before, block_field() would exit on getting any
field other than the existing fields
or default fields.
This allows getting other values,
like those that might have been added by an extending plugin.
For example, values added in the Inspector.

lukecarbis reviewed (approved):
Nice work here. Great test write up as well.

lukecarbis commented on php/helpers.php:
```diff
@@ -25,13 +25,27 @@ function block_field( $name, $echo = true ) {
-	$default_fields = array( 'className' );
+	$is_name_allowed = isset( $block_lab_config->fields[ $name ] ) || in_array( $name, $default_fields, true );
+
+	/**
```
Should this be a single asterix, since it's an inline comment, not a doc block?

i.e.

```
/*
 * Filters whether the $name argument to this function is allowed.
 *
 * …
 */
```

kienstra replied:
Ah, good question. 

The examples I've seen [in Core](https://github.com/WordPress/wordpress-develop/blob/00b03f2a6f2c21ebf9975c9c08976358681400ba/src/wp-includes/feed.php#L28) have had 2 asterisks, but I'm not sure if there's an official convention for this.

lukecarbis commented on php/helpers.php:
```diff
@@ -25,13 +25,27 @@ function block_field( $name, $echo = true ) {
+	 * @param bool   $is_name_allowed Whether the name is allowed.
+	 * @param string $name The name of value to get.
+	 */
+	$is_block_field_name_allowed = apply_filters( 'is_block_field_name_allowed', $is_name_allowed, $name );
```
Why are we filtering the `is_name_allowed` result, instead of filtering the `default_fields` array?

kienstra replied:
Ah, that's a good idea. I didn't think of that, as the `$default_fields` array didn't exist when I originally wrote this filter in another branch, about a month ago.

lukecarbis replied:
Ha! I only picked up on this because I thought you pulled me up on this a couple weeks back. :P

lukecarbis replied:
Maybe just do whatever we are already doing elsewhere in the plugin.

kienstra replied:
You make a good point, I think it's simpler to filter the `$default_fields`.

kienstra replied:
And thanks to the simplicity, there's less risk of regression.

kienstra committed:
Filter the $default_fields instead of a boolean

As Luke mentioned,
this makes more sense.
Typically, they would need to check the field name anyway.

kienstra committed:
Update the filter name in the unit tests

I had forgotten to do this in a few places,
including the tearDown() method.

kienstra commented:
Jeffrey should probably also review and test this before we merge it.

kienstra committed:
Update comments in tests, to refer to the current filter

Before, these comments referred to a previous
filter for whether the name is allowed.

lukecarbis commented:
@phpbits Could you take a look at this PR please?

kienstra commented:
@phpbits, sorry for the delay on this. The filter changed a little since we [discussed it](https://github.com/phpbits/block-options/blob/57a752f703590ea719bb9c85fad1f3a0e97b1fa0/includes/helper.php#L10-L17).

Here's how to use it now:

```php

/**
 * Adds 'editorskit' to the default fields, allowing it to be passed to block_field().
 *
 * @param array $default_fields The default fields.
 */
add_filter( 'block_lab_default_fields', function( $default_fields ) {
	$default_fields[] = 'editorskit';
	return $default_fields;
} );

```

phpbits commented:
@kienstra the new one doesn't work but the filter I've tested before is working perfectly. The problem is that `block_field( 'editorskit', false )` returns string `Array` instead of the array values. Thanks!

kienstra commented:
Hi @phpbits,
Thanks for testing this.

You're right, it looks like it's returning the string `'Array'`:

https://github.com/getblocklab/block-lab/blob/60e63eb6549a7fa1693d64e619278ae422de4b4f/php/helpers.php#L51-L53

I'll keep working on this.

phpbits commented:
Thanks a lot @kienstra!

kienstra committed:
Revert "Update comments in tests, to refer to the current filter"

This reverts commit 9bb5ce07a67e9e37af3c2a808d8209c20700e2f2.

kienstra committed:
Revert "Update the filter name in the unit tests"

This reverts commit 6714a2281a8a98791bfacbfb27f36a05e3d61d5a.

kienstra committed:
Revert "Filter the $default_fields instead of a boolean"

This reverts commit 99cf9797c923a8e2f322dda8113ac053e5630610.

kienstra committed:
Merge branch 'develop' into add/block-field-filter

kienstra commented:
**Question About Approach**

Hi @lukecarbis,
Do you think the current state of this PR is alright?

The idea you had of filtering the `$default_fields` was really good.

But it looks like there's something further down the function that prevents this from working. I only noticed that when Jeffrey brought this up.

In the case of Editorskit, it would add a `'editorskit'` to the `$default_fields` via a filter.

But further down in the function, it casts all `$default_fields` into a string, even though `editorskit` is an array:

https://github.com/getblocklab/block-lab/blob/60e63eb6549a7fa1693d64e619278ae422de4b4f/php/helpers.php#L51-L53

Maybe there's a better way to approach this, though.

I reverted all of the commits to this PR, except [the first](https://github.com/getblocklab/block-lab/pull/412/commits/77d01f641e8a865dfd0d65eeaf034a94895482cd).

It'd be good to have this PR merged in the next release, but only if we get this approach right, and Jeffrey gets a chance to test it.

Thanks, Luke!

lukecarbis commented:
@kienstra I must be missing something – where is `$default_fields` cast as a string?

lukecarbis commented:
@kienstra Would it work if we instantiated each of the default fields as a `Block_Lab\Blocks\Field`? That way we could call `cast_value` against it, instead of forcing it to be a string.

e.g.
```PHP
$default_fields = apply_filters(
    'block_lab_default_fields',
   array(
       'className' => new Field(
            array(
                'name' => 'className',
                'type' => 'string'
            )
        )
    )
);
```

and later…

```PHP
} elseif ( array_key_exists( $name, $default_fields, true ) ) { 
    // Cast default Editor attributes appropriately. 
    $value = $default_fields[ $name ]->cast_value( $value ); 
}
```

kienstra committed:
Reinstate earlier edits that change filter to 'block_lab_default_fields'

Reverts the last 3 'revert' commits.
To try a new approach.

kienstra committed:
Thanks to Luke's suggestion, instantiate a Field object to run ->cast_value()

Tries an approach of instantiating a Field
for any field passed to a filter.

kienstra commented:
Hi @lukecarbis,
That's a good idea.

The last 2 commits have a slightly modified version of your idea. The `block_field()` method takes care of instantiating a `Field`, as an external plugin would have to do `new Block_Lab\Blocks\Field()`.

Now, adding a `$default_field` would look like:

```php
/**
 * Adds 'editorskit' to the default fields, allowing it to be passed to block_field().
 *
 * @param array $default_fields The default fields.
 */
add_filter( 'block_lab_default_fields', function( $default_fields ) {
	$default_fields['editorskit'] = 'array'; // The name and type.

	return $default_fields;
} );
```

kienstra commented on php/helpers.php:
```diff
@@ -29,9 +29,22 @@ function block_field( $name, $echo = true ) {
+	 * But this allows getting block attributes that might have been added by other plugins.
+	 * To allow getting another attribute, add it to the $default_fields array.
+	 *
+	 * @param array  $default_fields An associative array of $field_name => $field_type.
```
This could probably use more description of the types that are possible, like string, array, etc...

lukecarbis reviewed (approved):
These changes look great, @kienstra! Nice work. Please merge whenever you and @phpbits are ready.

kienstra commented:
**Request To Test**

Hi @phpbits,
Sorry to keep changing this, and hope you've been doing great.

When you have a chance, could you please test this PR?

Here's how to use the new filter:

```php
/**
 * Adds 'editorskit' to the default fields, allowing it to be passed to block_field().
 *
 * @param array $default_fields An associative array of $field_name => $field_type.
 */
add_filter( 'block_lab_default_fields', function( $default_fields ) {
	$default_fields['editorskit'] = 'array'; // The name and type.

	return $default_fields;
} );
```

phpbits commented:
@kienstra My apology for late response. Just tested it and working perfectly 💯 Thanks a lot!

kienstra commented:
Thanks, @phpbits! Sorry for all of the changes here.

kienstra committed:
Slightly improve the DocBlock for the filter

Add a line with an example of how to use the filter.
Though this doesn't touch the actual logic.

kienstra commented:
**Going To Merge If That's Alright**

Hi @lukecarbis,
If it's alright, I'm going to merge this soon. 

@phpbits kindly tested this again, and I made a [comment-only commit](https://github.com/getblocklab/block-lab/pull/412/commits/669439f463e4ef2177b2b268549766e0a76449c4) to make the DocBlock a little more clear.

It'd be good to have this in production soon.

Thanks, Luke!
